src/core/surface/completion_queue.c

```diff
@@ -427,7 +439,19 @@ grpc_event grpc_completion_queue_pluck(grpc_completion_queue *cc, void *tag,
       break;
     }
     first_loop = 0;
-    grpc_pollset_work(&exec_ctx, &cc->pollset, &worker, now, deadline);
+    /* Check alarms - these are a global resource so we just ping
+       each time through on every pollset.
+       May update deadline to ensure timely wakeups.
+       TODO(ctiller): can this work be localized? */
+    gpr_timespec iteration_deadline = deadline;
+    if (grpc_timer_check(&exec_ctx, now, &iteration_deadline)) {
+      GPR_TIMER_MARK("alarm_triggered", 0);
+      gpr_mu_unlock(GRPC_POLLSET_MU(&cc->pollset));
+      grpc_exec_ctx_flush(&exec_ctx);
+      gpr_mu_lock(GRPC_POLLSET_MU(&cc->pollset));
+      continue;
+    }
+    grpc_pollset_work(&exec_ctx, &cc->pollset, &worker, now, iteration_deadline);
```

Comment:
If we reuse the code we end up in the same complicated tracking mess - I
like the fact that the mutex manipulation is mostly surfaced now.

On Tue, Feb 16, 2016 at 9:54 AM Vijay Pai <notifications@github.com> wrote:

> In src/core/surface/completion_queue.c
> <https://github.com/grpc/grpc/pull/5257#discussion_r53049561>:
>
> > @@ -427,7 +439,19 @@ grpc_event grpc_completion_queue_pluck(grpc_completion_queue *cc, void *tag,
> >        break;
> >      }
> >      first_loop = 0;
> > -    grpc_pollset_work(&exec_ctx, &cc->pollset, &worker, now, deadline);
> > +    /* Check alarms - these are a global resource so we just ping
> > +       each time through on every pollset.
> > +       May update deadline to ensure timely wakeups.
> > +       TODO(ctiller): can this work be localized? */
> > +    gpr_timespec iteration_deadline = deadline;
> > +    if (grpc_timer_check(&exec_ctx, now, &iteration_deadline)) {
> > +      GPR_TIMER_MARK("alarm_triggered", 0);
> > +      gpr_mu_unlock(GRPC_POLLSET_MU(&cc->pollset));
> > +      grpc_exec_ctx_flush(&exec_ctx);
> > +      gpr_mu_lock(GRPC_POLLSET_MU(&cc->pollset));
> > +      continue;
> > +    }
> > +    grpc_pollset_work(&exec_ctx, &cc->pollset, &worker, now, iteration_deadline);
>
> Any way to reuse code here since it's the same as the above case?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/grpc/grpc/pull/5257/files#r53049561>.
>